modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/SearchTemplateRequestTests.java

```diff
@@ -19,24 +19,179 @@
+        request.setProfile(true);
+
+        Map<String, Object> scriptParams = new HashMap<>();
+        scriptParams.put("my_field", "foo");
+        scriptParams.put("my_value", "bar");
+        request.setScriptParams(scriptParams);
+
+        XContentType contentType = randomFrom(XContentType.values());
+        XContentBuilder expectedRequest = XContentFactory.contentBuilder(contentType)
+            .startObject()
+                .field("source", "{\"query\": { \"match\" : { \"{{my_field}}\" : \"{{my_value}}\" } } }")
+                .startObject("params")
+                    .field("my_field", "foo")
+                    .field("my_value", "bar")
+                .endObject()
+                .field("explain", false)
+                .field("profile", true)
+            .endObject();
+
+        XContentBuilder actualRequest = XContentFactory.contentBuilder(contentType);
+        request.toXContent(actualRequest, ToXContent.EMPTY_PARAMS);
+
+        assertToXContentEquivalent(BytesReference.bytes(expectedRequest),
+            BytesReference.bytes(actualRequest),
+            contentType);
+    }
+
+    public void testToXContentWithStoredTemplate() throws IOException {
+        SearchTemplateRequest request = new SearchTemplateRequest();
+
+        request.setScriptType(ScriptType.STORED);
+        request.setScript("match_template");
+        request.setExplain(true);
+
+        Map<String, Object> params = new HashMap<>();
+        params.put("my_field", "foo");
+        params.put("my_value", "bar");
+        request.setScriptParams(params);
+
+        XContentType contentType = randomFrom(XContentType.values());
+        XContentBuilder expectedRequest = XContentFactory.contentBuilder(contentType)
+            .startObject()
+                .field("id", "match_template")
+                .startObject("params")
+                    .field("my_field", "foo")
+                    .field("my_value", "bar")
+                .endObject()
+                .field("explain", true)
+                .field("profile", false)
+            .endObject();
+
+        XContentBuilder actualRequest = XContentFactory.contentBuilder(contentType);
+        request.toXContent(actualRequest, ToXContent.EMPTY_PARAMS);
+
+        assertToXContentEquivalent(
+            BytesReference.bytes(expectedRequest),
+            BytesReference.bytes(actualRequest),
+            contentType);
+    }
+
+    /**
+     * Note that for xContent parsing, we omit two parts of the request:
+     * - The 'simulate' option is always held constant, since this parameter is not included
+     *   in the request's xContent (it's instead used to determine the request endpoint).
+     * - We omit the random SearchRequest, since this component only affects the request
+     *   parameters and also isn't captured in the request's xContent.
+     */
+    public void testFromXContent() throws IOException {
```

Comment:
Given these changes, I wonder if we should rather extend `AbstractStreamableTestCase`. We've had these problems in other cases, requests tend to require different equal comparisons depending on whether we are testing transport serialization or xcontent serialization. Transport serialization includes request parameters that are not part of the xcontent serialization, which is only about the http request body after all.
Got the xcontent part, maybe we could have a separate test class that extends `AbstractXContentTestCase`, where you can override `assertEqualInstances` and tweak all you need without having to override testFromXContent (hopefully)?

Reply:
I ended up splitting out the xContent parts of this test into a new class, `SearchTemplateRequestXContentTests`. Please let me know if this is not what you had in mind.